Issue 13: Wrong XML? not fine, not fine

DiegoPino opened the pull request:
# What does this pull solve?
See #13 and also #12 (hopefully)

Basically it checks if the resulting find and replace operation against a given XML data stream ends in a valid (simply valid in terms of an actual XML that can be parsed, we are not validating if its MODS, etc). I checked the whole islandora code base and there IS NOT A SINGLE check anywhere when dealing with XML. Its terrible. All depends on if Fedora takes or not the file and if it dies/throws errors or not. So disappointed..

How does this work?

1.- We check first here if there is an actual change. I mean why do we need to update a data stream if nothing is going to be changed?
2.- if that passes, we check the mime-type of the data stream, if its text/xml, application/xml or text/html we simply, silenty load it via the simple xml class. If it fails we assume (correctly) it's wrong. This applies to things like "hey" i want to add an `&` or a  on closing tag. 

3.- If it fails we add a new key to the batch context named  "total failure" and assume that one bad replacement will/could lead to the same in the rest and stop processing but skipping further attempts. that is the way. If your pattern can lead to wrong XML then change it.

How to test? Git checkout, git pull, choose a collection you care little about, try find any closing tag and replace it by an "&". Do it, should start and finish quick and give you a message and a list of failed replacements that has all the ones that matched originally in the list. Also there will be in the Reports/logs entries about the error with a link to that find and replace log id.

Finally it moves the namespace check to the end of the filter chain. Have not tested that but at least? it does not fail. Can anyone see if it is working fine?

@bondjimbond @flummingbird maybe there is interest there to test this? Not sure how many of you are still using this, but this makes this module wayyyyyyy safer.

DiegoPino committed:
First pass on checking valid XML and aborting batch if one is wrong

Because if one is wrong, many others could be! Better be safe and make the repo work less and the user be more engaged

DiegoPino committed:
Add totalfailure flag

This won't kill the from calling the operation but will simply skip them but keep the log around so we know how many were affected. This is just for consistency really.

DiegoPino committed:
Add a link to the ID Log.

DiegoPino committed:
Wrong relative path for the link

This should be OK now!

DiegoPino commented:
Ok, the filter/conditional order seems to be Ok? Not sure if it would work for the use case of "no cmodel"

E.g this is an SPARQL for just collection, CMODEL and namespace (that works). 
```
PREFIX fm: <info:fedora/fedora-system:def/model#>
            PREFIX fr: <info:fedora/fedora-system:def/relations-external#>
            PREFIX fv: <info:fedora/fedora-system:def/view#>
            PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
            SELECT ?object ?label
            FROM <#ri>
            WHERE {
              ?object fm:hasModel <info:fedora/islandora:sp_large_image_cmodel> ; fr:isMemberOfCollection <info:fedora/hamilton:taylor> ; FILTER (REGEX( STR(?object), "^info:fedora/hamilton:", "i"))
              OPTIONAL{
                ?object fm:label ?label
              }
           }
```

This one with all also works
```
PREFIX fm: <info:fedora/fedora-system:def/model#>
            PREFIX fr: <info:fedora/fedora-system:def/relations-external#>
            PREFIX fv: <info:fedora/fedora-system:def/view#>
            PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
            SELECT ?object ?label
            FROM <#ri>
            WHERE {
              ?object fm:hasModel <info:fedora/islandora:sp_large_image_cmodel> ; fr:isMemberOfCollection <info:fedora/hamilton:taylor> ; fv:lastModifiedDate ?date
      FILTER ( ?date >= xsd:dateTime("2010-08-18T00:00:00Z") && ?date <= xsd:dateTime("2020-08-18T23:59:59Z"))FILTER (REGEX( STR(?object), "^info:fedora/hamilton:", "i"))
              OPTIONAL{
                ?object fm:label ?label
              }
           }
```

bondjimbond commented:
Error when I try to run a basic find/replace before even testing the validation.

1. Content model: PDF (I have one PDF in my test environment)
2. DSID = MODS
3. Search text = `<dateIssued>`
4. Replace text = `<dateIssued> `

Result:

```
PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'isregex' in 'field list': INSERT INTO {islandora_find_replace} (state, cmodel, dsid, find, replacement, isregex) VALUES (:db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2, :db_insert_placeholder_3, :db_insert_placeholder_4, :db_insert_placeholder_5); Array ( [:db_insert_placeholder_0] => submitted [:db_insert_placeholder_1] => islandora:sp_pdf [:db_insert_placeholder_2] => MODS [:db_insert_placeholder_3] => <dateIssued> [:db_insert_placeholder_4] => <dateIssued [:db_insert_placeholder_5] => 0 ) in drupal_write_record() (line 7457 of /var/www/drupal/includes/common.inc).
```

DiegoPino commented:
Wrong branch maybe? That column has been around for some time. Maybe a
drush updatedb missing?

El El mié, 19 de ago. de 2020 a la(s) 10:12, Brandon Weigel <
notifications@github.com> escribió:

>
>
> Error when I try to run a basic find/replace before even testing the
> validation.
>
>
>
>
>
>    1. Content model: PDF (I have one PDF in my test environment)
>
>    2. DSID = MODS
>
>    3. Search text = <dateIssued>
>
>    4. Replace text = <dateIssued>
>
>
>
>
> Result:
>
>
> PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'isregex' in 'field list': INSERT INTO {islandora_find_replace} (state, cmodel, dsid, find, replacement, isregex) VALUES (:db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2, :db_insert_placeholder_3, :db_insert_placeholder_4, :db_insert_placeholder_5); Array ( [:db_insert_placeholder_0] => submitted [:db_insert_placeholder_1] => islandora:sp_pdf [:db_insert_placeholder_2] => MODS [:db_insert_placeholder_3] => <dateIssued> [:db_insert_placeholder_4] => <dateIssued [:db_insert_placeholder_5] => 0 ) in drupal_write_record() (line 7457 of /var/www/drupal/includes/common.inc).
>
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mnylc/islandora_find_replace/pull/14#issuecomment-676423475>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU7ZZ5URRFRR4SFIE4MG63SBPMTLANCNFSM4QEAPIPA>
> .
>
>
> --
Diego Pino Navarro
Digital Repositories Developer
Metropolitan New York Library Council (METRO)

bondjimbond commented:
Ah, yes, needed to run update.php, thanks.

So I tried to create some invalid XML, and it did stop me from doing so, but there are some errors showing at the same time:

Search text: `<dateIssued>`
Replacement text: `<dateIssued`

Result:

 > This find and Replaced Batch update was suspended because one replace operation leads to an invalid XML. Please check your logs and or/replace your search & replacement patterns.

> Find and Replace complete!

```
Error message
Notice: Undefined index: success in islandora_find_replace_operation_log() (line 37 of /var/www/drupal/sites/all/modules/islandora_find_replace/includes/log.inc).
Warning: Invalid argument supplied for foreach() in islandora_find_replace_operation_log() (line 37 of /var/www/drupal/sites/all/modules/islandora_find_replace/includes/log.inc).
```

A couple of things:

- The "batch suspended" notice should probably not be inside a green checkmark box, since it's misleading (it's showing you an error, not a success)

- Since it's suspended, it should not show "Find and replace complete!"

- Looks like there are some logging errors happening because the batch was interrupted

DiegoPino commented:
Ok, good. Some cosmetics and i need to add a check on the logs for that new
state (which is neither full error not success) . Will resolve once back in
the office. Thanks!

DiegoPino committed:
Address code review

- Checks if there are actuall replace results before iterating over them (that was an existing but from old times)
- Changes Status message to Warning message (not error because we caught the error by suspending the operation)
- Do not show Find and replace something exclamation point if replace ended somewhere in an issue.

bondjimbond commented:
Hey, it's good! Thanks for this fix.

bondjimbond reviewed (approved):
I'm not a committer, but I approve.

DiegoPino commented:
@bondjimbond you are a committer here man and your opinion is binding. Great! Merging
